### DISCUSSION TITLE
ESLint 9

### DISCUSSION THREAD

**Discussion by lishaduck** [AndreaPontrandolfo/sheriff#155] [General]
Last one 😅
Is there a particular reason/dependency holding Sheriff back from ESLint 9 compatibility?
What needs to be done to migrate?

**Comment by AndreaPontrandolfo** [comment]
I would love to migrate Sheriff to V9.
I'm still exploring here, but it seems like a daunting task.
Most of the ecosystem is still on V8, with some big maintainers saying that the migration will take them months, if not years.
ESLint made a lot of API breaking changes with V9.
I think a lot plugins would break, but i didn't even try.

I think a quick thing we could try is:
- migrate `typescript-eslint` to latest, because the latest version is V9-ready
- migrate `eslint-plugin-unicorn` to latest, because the latest version is V9-ready
- wrap most (if not all) plugins that we are using with [compat](https://github.com/eslint/rewrite/tree/main/packages/compat)
- hope for the best lol

**Reply by lishaduck** [reply]
Yeah. tseslint support is certainly needed first, and stylistic. (both in progress)
Unicorn supports 9, as you said.
@eslint/js obviously supports 9
eslint-config-prettier doesn't support 9, nor does it support stylistic
eslint-plugin-react{,-jsx} doesn't plan on support eslint 9 anytime soon, it seems, at it means at least dropping eslint 3
eslint-plugin-react-hooks supports 9
eslint-plugin-react-refresh seems not to have broken, so implicit support? (no comments about 9, but docs for flat config)
@eslint-react/eslint-plugin only supports 9
eslint-plugin-sonarjs supports 9
eslint-plugin-fp obviously doesn't support 9, it doesn't even support 8 officially (no issues about 9 at all)
eslint-plugin-jsdoc only supports 9
eslint-plugin-tsdoc is basically stable, and therefore basically dead (only 8, no replies to update issues)
eslint-plugin-jest supports flat config, maybe 9?
eslint-plugin-vitest supports 9 (maybe unreleased?)
eslint-plugin-import doesn't support 9 (import-x works with the compat utils, so maybe time to switch?)
ESLint-Plugin-Lodash-f (@AndreaPontrandolfo, comment?)
eslint-plugin-next supports 7 & 8, but not 9
eslint-plugin-playwright supports flat config, 9 unknown
eslint-plugin-astro supports both 8 and 9
eslint-plugin-fsecond (@AndreaPontrandolfo, comment? Oh, and there's a typo in your readme. 3 rules now, not 2)
eslint-plugin-prefer-early-return is unknown
eslint-plugin-arrow-return-style likely does not support 9
I actually have eslint-plugin-simple-import-sort open, I can check if it works.

Obviously they all support flat config, as we use them, but noting if they've bothered to begin looking into support for 9 features.

**Reply by AndreaPontrandolfo** [reply]
[eslint-plugin-react now supports V9](https://github.com/jsx-eslint/eslint-plugin-react/releases/tag/v7.35.0)!

**Reply by lishaduck** [reply]
Yay!

P.S.: I think I'm write up a gist for this.

**Reply by lishaduck** [reply]
> I think I'm write up a gist for this.

I began, but then discovered eslint/eslint#18093 & eslint/eslint#18391.

**Reply by AndreaPontrandolfo** [reply]
Do you see an issue in particular?

**Reply by lishaduck** [reply]
I think `import` & `import-x` support is blocking. Otherwise, no.
Oh, maybe prettier?

**Reply by AndreaPontrandolfo** [reply]
I _think_ that `import-x` is already compatible with V9, while `import` is probably not compatible but the only breaking rule should be `no-unused-modules`, which we specifically don't use (as noted in the [FAQS](https://www.eslint-config-sheriff.dev/docs/faq#why-you-didnt-include-eslint-pluginsrules-for-x-library))

**Reply by lishaduck** [reply]
> I _think_ that `import-x` is already compatible with V9, while `import` is probably not compatible but the only breaking rule should be `no-unused-modules`, which we specifically don't use (as noted in the [FAQS](https://www.eslint-config-sheriff.dev/docs/faq#why-you-didnt-include-eslint-pluginsrules-for-x-library))

Yeah, I guess you're right about that. I think a PR for ESLint 9 might be coming ~soon (in a month, if you haven't done so yourself[^1])!

[^1]: I have my Mac through my school, and I have to return it. I'll get it back mid-August, but until then I'll be on my Windows machine, and it's rather slow and isn't full of space. I can't spend 5 or 6 gigs of space on OSS contributions. Sorry 🙃 Be back ~~soon~~.

**Comment by AndreaPontrandolfo** [comment]
Tracking milestone: https://github.com/AndreaPontrandolfo/sheriff/milestone/1

**Comment by AndreaPontrandolfo** [comment]
[Related issue](https://github.com/AndreaPontrandolfo/sheriff/issues/185)

**Comment by AndreaPontrandolfo** [comment]
Full support for ESLint v9 landed with [release 20](https://github.com/AndreaPontrandolfo/sheriff/releases/tag/eslint-config-sheriff%4020.0.0).
As time goes goes on, we will probably remove all `fixupPluginRules()` call instances, gradually. But that's a minor detail.

This discussion is now resolved.
